---
title: sort-named-exports
---

# sort-named-exports

ğŸ’¼ This rule is enabled in the following [configs](/configs/): `recommended-alphabetical`, `recommended-line-length`, `recommended-natural`.

ğŸ”§ This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

## ğŸ“– Rule Details

Enforce sorted named exports.

Maintaining a consistent and sorted order of named exports can improve code readability.

## ğŸ’¡ Examples

### Alphabetical and Natural Sorting

<!-- prettier-ignore -->
```ts
// Incorrect
export {
  Emma,
  Ray,
  Norman,
  Don,
  Gilda,
} from 'grace-field'

// Correct
export {
  Don,
  Emma,
  Gilda,
  Norman,
  Ray,
} from 'grace-field'
```

### Sorting by Line Length

<!-- prettier-ignore -->
```ts
// Incorrect
export {
  Emma,
  Ray,
  Norman,
  Don,
  Gilda,
} from 'grace-field'

// Correct
export {
  Norman,
  Gilda,
  Emma,
  Don,
  Ray,
} from 'grace-field'
```

## ğŸ”§ Options

### `type`

- `enum` (default: `alphabetical`):
  - `alphabetical` - sort alphabetically.
  - `natural` - sort in natural order.
  - `line-length` - sort by code line length.

### `order`

- `enum` (default: `asc`):
  - `asc` - enforce properties to be in ascending order.
  - `desc` - enforce properties to be in descending order.

## âš™ï¸ Usage

### Legacy Config

```json
// .eslintrc
{
  "rules": {
    "perfectionist/sort-named-exports": [
      "error",
      {
        "type": "line-length",
        "order": "desc"
      }
    ]
  }
}
```

### Flat Config

```js
// eslint.config.js
import perfectionist from 'eslint-plugin-perfectionist'

export default [
  {
    plugins: {
      perfectionist,
    },
    rules: {
      'perfectionist/sort-named-exports': [
        'error',
        {
          type: 'line-length',
          order: 'desc',
        },
      ],
    },
  },
]
```

## ğŸš€ Version

This rule was introduced in v0.4.0.

## ğŸ“š Resources

- [Rule source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/rules/sort-named-exports.ts)
- [Test source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/test/sort-named-exports.test.ts)
