---
title: sort-interfaces
description: ESLint Plugin Perfectionist rule which enforce sorted TypeScript interface properties
---

# sort-interfaces

üíº This rule is enabled in the following [configs](/configs/): `recommended-alphabetical`, `recommended-line-length`, `recommended-natural`.

üîß This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).

<!-- end auto-generated rule header -->

## üìñ Rule Details

Enforce sorted TypeScript interface properties.

Sorting interface properties provides a clear and predictable structure to the codebase, making it easier for developers to locate the various properties defined within an interface. It helps maintain consistency and allows for efficient maintenance and collaboration among team members.

It's **safe**. If you document interface properties line by line, property comments will also be sorted.

:::info Important
If you use the [`adjacent-overload-signatures`](https://typescript-eslint.io/rules/adjacent-overload-signatures) rule from the [`@typescript-eslint/eslint-plugin`](https://typescript-eslint.io) plugin, it is highly recommended to [disable it](https://eslint.org/docs/latest/use/configure/rules#using-configuration-files-1) to avoid conflicts.
:::

## üí° Examples

::: code-group

<!-- prettier-ignore -->
```ts [Alphabetical and Natural Sorting]
// ‚ùå Incorrect
interface ButtonProps {
  variant?: 'solid' | 'outline' | 'text'
  full?: boolean
  disabled?: ComponentProps<'button'>['disabled']
  type?: 'submit' | 'button'
  color: 'main' | 'info' | 'success' | 'warning' | 'error'
  onClick?: () => void
  children?: string | number
  icon?: FC<SVGProps<SVGSVGElement>>
  className?: string
  size: 's' | 'm' | 'l'
}

// ‚úÖ Correct
interface ButtonProps {
  children?: string | number
  className?: string
  color: 'main' | 'info' | 'success' | 'warning' | 'error'
  disabled?: ComponentProps<'button'>['disabled']
  full?: boolean
  icon?: FC<SVGProps<SVGSVGElement>>
  onClick?: () => void
  size: 's' | 'm' | 'l'
  type?: 'submit' | 'button'
  variant?: 'solid' | 'outline' | 'text'
}
```

```ts [Sorting by Line Length]
// ‚ùå Incorrect
interface ButtonProps {
  variant?: 'solid' | 'outline' | 'text'
  full?: boolean
  disabled?: ComponentProps<'button'>['disabled']
  type?: 'submit' | 'button'
  color: 'main' | 'info' | 'success' | 'warning' | 'error'
  onClick?: () => void
  children?: string | number
  icon?: FC<SVGProps<SVGSVGElement>>
  className?: string
  size: 's' | 'm' | 'l'
}

// ‚úÖ Correct
interface ButtonProps {
  color: 'main' | 'info' | 'success' | 'warning' | 'error'
  disabled?: ComponentProps<'button'>['disabled']
  variant?: 'solid' | 'outline' | 'text'
  icon?: FC<SVGProps<SVGSVGElement>>
  children?: string | number
  type?: 'submit' | 'button'
  size: 's' | 'm' | 'l'
  onClick?: () => void
  className?: string
  full?: boolean
}
```

:::

## üîß Options

This rule accepts an options object with the following properties:

```ts
type CustomGroup = string
type Group = 'multiline' | CustomGroup

interface Options {
  type?: 'alphabetical' | 'natural' | 'line-length'
  'optionality-order'?: 'ignore' | 'optional-first' | 'required-first'
  order?: 'asc' | 'desc'
  'ignore-case'?: boolean
  groups?: (Group | Group[])[]
  'custom-groups'?: { [key: Group]: string[] | string }
  'ignore-pattern'?: string[]
  'partition-by-new-line'?: boolean
}
```

### type

<sub>(default: `'alphabetical'`)</sub>

- `alphabetical` - sort alphabetically.
- `natural` - sort in natural order.
- `line-length` - sort by code line length.

### optionality-order

<sub>(default: `'ignore'`)</sub>

- `optional-first` - put all optional members first.
- `required-first` - put all required members first.

### order

<sub>(default: `'asc'`)</sub>

- `asc` - enforce properties to be in ascending order.
- `desc` - enforce properties to be in descending order.

### ignore-case

<sub>(default: `false`)</sub>

Only affects alphabetical and natural sorting. When `true` the rule ignores the case-sensitivity of the order.

### groups

<sub>(default: `[]`)</sub>

You can set up a list of interface groups for sorting. Groups can be combined. There are predefined group: `'multiline'`.

### custom-groups

<sub>(default: `{}`)</sub>

You can define your own groups for object keys. The [minimatch](https://github.com/isaacs/minimatch) library is used for pattern matching.

Example:

```
{
  "custom-groups": {
    "top": "id"
  }
}
```

### ignore-pattern

<sub>(default: `[]`)</sub>

If you need to ignore a rule for some interfaces, you can specify their names or a pattern to ignore, for example: `'Component*'` to ignore all interfaces whose names begin with the word Component.

The [minimatch](https://github.com/isaacs/minimatch) library is used for pattern matching.

### partition-by-new-line

<sub>(default: `false`)</sub>

When `true`, does not sort the interface's element if there is an empty string between them.

## ‚öôÔ∏è Usage

::: code-group

```json [Legacy Config]
// .eslintrc
{
  "plugins": ["perfectionist"],
  "rules": {
    "perfectionist/sort-interfaces": [
      "error",
      {
        "type": "natural",
        "order": "asc"
      }
    ]
  }
}
```

```js [Flat Config]
// eslint.config.js
import perfectionist from 'eslint-plugin-perfectionist'

export default [
  {
    plugins: {
      perfectionist,
    },
    rules: {
      'perfectionist/sort-interfaces': [
        'error',
        {
          type: 'natural',
          order: 'asc',
        },
      ],
    },
  },
]
```

:::

## üöÄ Version

This rule was introduced in v0.1.0.

## üìö Resources

- [Rule source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/rules/sort-interfaces.ts)
- [Test source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/test/sort-interfaces.test.ts)
