---
title: sort-exports
description: Enforce a consistent order of exports to enhance code readability and maintainability. Use this ESLint rule to keep your exports well-organized
shortDescription: Enforce sorted exports
keywords:
  - eslint
  - sort exports
  - eslint rule
  - coding standards
  - code quality
  - javascript linting
  - module exports sorting
---

import CodeExample from '../../components/CodeExample.svelte'
import CodeTabs from '../../components/CodeTabs.svelte'
import { dedent } from 'ts-dedent'

Enforce sorted exports.

Sorting exports in a consistent manner can greatly enhance the readability and maintainability of your codebase. By alphabetically ordering exports, developers can quickly identify and correct any missing or incorrect exports.

This rule promotes a structured approach to exporting modules, which simplifies code management and improves the overall quality of the project.

## Try it out

<CodeExample
  alphabetical={dedent`
    export { createUser } from './actions/create-user'
    export { deleteUser } from './actions/delete-user'
    export { fetchUser } from './actions/fetch-user'
    export { updateUser } from './actions/update-user'
    export { Footer } from './components/Footer'
    export { Header } from './components/Header'
    export { MainContent } from './components/MainContent'
    export { Sidebar } from './components/Sidebar'
    export { calculateAge } from './utils/calculate-age'
    export { debounce } from './utils/debounce'
    export { formatDate } from './utils/format-date'
    export { generateUUID } from './utils/generate-uuid'
    export { parseQueryString } from './utils/parse-query-string'
  `}
  lineLength={dedent`
    export { parseQueryString } from './utils/parse-query-string'
    export { MainContent } from './components/MainContent'
    export { calculateAge } from './utils/calculate-age'
    export { generateUUID } from './utils/generate-uuid'
    export { createUser } from './actions/create-user'
    export { deleteUser } from './actions/delete-user'
    export { updateUser } from './actions/update-user'
    export { fetchUser } from './actions/fetch-user'
    export { formatDate } from './utils/format-date'
    export { Sidebar } from './components/Sidebar'
    export { Footer } from './components/Footer'
    export { Header } from './components/Header'
    export { debounce } from './utils/debounce'
  `}
  initial={dedent`
    export { MainContent } from './components/MainContent'
    export { calculateAge } from './utils/calculateAge'
    export { Sidebar } from './components/Sidebar'
    export { deleteUser } from './actions/deleteUser'
    export { Footer } from './components/Footer'
    export { debounce } from './utils/debounce'
    export { generateUUID } from './utils/generateUUID'
    export { formatDate } from './utils/formatDate'
    export { updateUser } from './actions/updateUser'
    export { fetchUser } from './actions/fetchUser'
    export { Header } from './components/Header'
    export { createUser } from './actions/createUser'
    export { parseQueryString } from './utils/parseQueryString'
  `}
  client:load
  lang="tsx"
/>

## Options

This rule accepts an options object with the following properties:

### type

<sub>default: `'alphabetical'`</sub>

Specifies the sorting method.

- `'alphabetical'` — Sort items alphabetically (e.g., “a” < “b” < “c”) using [localeCompare](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare).
- `'natural'` — Sort items in a [natural](https://github.com/yobacca/natural-orderby) order (e.g., “item2” < “item10”).
- `'line-length'` — Sort items by the length of the code line (shorter lines first).

### order

<sub>default: `'asc'`</sub>

Determines whether the sorted items should be in ascending or descending order.

- `'asc'` — Sort items in ascending order (A to Z, 1 to 9).
- `'desc'` — Sort items in descending order (Z to A, 9 to 1).

### ignoreCase

<sub>default: `true`</sub>

Controls whether sorting should be case-sensitive or not.

- `true` — Ignore case when sorting alphabetically or naturally (e.g., “A” and “a” are the same).
- `false` — Consider case when sorting (e.g., “a” comes before “A”).

### specialCharacters

<sub>default: `keep`</sub>

Controls whether special characters should be trimmed, removed or kept before sorting.

- `'keep'` — Keep special characters when sorting (e.g., “_a” comes before “a”).
- `'trim'` — Trim special characters when sorting alphabetically or naturally (e.g., “_a” and “a” are the same).
- `'remove'` — Remove special characters when sorting (e.g., “/a/b” and “ab” are the same).

### locales

<sub>default: `'en-US'`</sub>

Specifies the sorting locales. See [String.prototype.localeCompare() - locales](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare#locales).

- `string` — A BCP 47 language tag (e.g. `'en'`, `'en-US'`, `'zh-CN'`).
- `string[]` — An array of BCP 47 language tags.

**Note:** Only effective when `type` is `alphabetical`.

### partitionByComment

<sub>default: `false`</sub>

Allows you to use comments to separate the exports into logical groups. This can help in organizing and maintaining large export blocks by creating partitions within the enum based on comments.

- `true` — All comments will be treated as delimiters, creating partitions.
-	`false` — Comments will not be used as delimiters.
- `string` — A regexp pattern to specify which comments should act as delimiters.
- `string[]` — A list of regexp patterns to specify which comments should act as delimiters.

### partitionByNewLine

<sub>default: `false`</sub>

When `true`, the rule will not sort the exports if there is an empty line between them. This can be useful for keeping logically separated groups of exports in their defined order.

```js
// Group 1
export * from "./atoms";
export * from "./organisms";
export * from "./shared";

// Group 2
export { Named } from './folder';
export { AnotherNamed } from './second-folder';
```

Each group of exports (separated by empty lines) is treated independently, and the order within each group is preserved.

### groupKind

<sub>default: `'mixed'`</sub>

Allows you to group exports by their kind, determining whether value exports should come before or after type exports.

- `mixed` — Do not group named exports by their kind; export statements are sorted together regardless of their type.
- `values-first` — Group all value exports before type exports.
- `types-first` — Group all type exports before value exports.

## Usage

<CodeTabs
  code={[
    {
      source: dedent`
        // eslint.config.js
        import perfectionist from 'eslint-plugin-perfectionist'

        export default [
          {
            plugins: {
              perfectionist,
            },
            rules: {
              'perfectionist/sort-exports': [
                'error',
                {
                  type: 'alphabetical',
                  order: 'asc',
                  ignoreCase: true,
                  specialCharacters: 'keep',
                  partitionByComment: false,
                  partitionByNewLine: false,
                  groupKind: 'mixed',
                },
              ],
            },
          },
        ]
      `,
      name: 'Flat Config',
      value: 'flat',
    },
    {
      source: dedent`
        // .eslintrc.js
        module.exports = {
          plugins: [
            'perfectionist',
          ],
          rules: {
            'perfectionist/sort-exports': [
              'error',
              {
                type: 'alphabetical',
                order: 'asc',
                ignoreCase: true,
                specialCharacters: 'keep',
                partitionByComment: false,
                partitionByNewLine: false,
                groupKind: 'mixed',
              },
            ],
          },
        }
      `,
      name: 'Legacy Config',
      value: 'legacy',
    },
  ]}
  type="config-type"
  client:load
  lang="tsx"
/>


## Version

This rule was introduced in [v1.2.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v1.2.0).

## Resources

- [Rule source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/rules/sort-exports.ts)
- [Test source](https://github.com/azat-io/eslint-plugin-perfectionist/blob/main/test/sort-exports.test.ts)
